syntax = "proto3";

package tgbot.builder.android;

import "google/protobuf/empty.proto";

// Settings for Android ROM builds
// do_repo_sync: Whether to sync the repo before building
// do_clean_build: Whether to perform a clean build
// use_ccache: Whether to use ccache for builds
// use_rbe_service: Whether to use Remote Build Execution service
// rbe_api_token: API token for RBE service authentication, if not set, RBE will not be used.
// do_upload: Whether to upload the built ROM after completion
message Settings {
    optional bool do_repo_sync = 1;
    optional bool do_clean_build = 2;
    optional bool use_ccache = 3;
    optional bool use_rbe_service = 4;
    optional string rbe_api_token = 5;
    optional bool do_upload = 6;
}

enum CleanDirectoryType {
    ROMDirectory = 0;
    BuildDirectory = 1;
}

// Request to clean a directory
message CleanDirectoryRequest {
    CleanDirectoryType directory_type = 1;
}

// Build variants for Android ROMs
enum BuildVariant {
    User = 0;
    UserDebug = 1;
    Eng = 2;
}

enum UploadMethod {
    None = 0;      // No upload
    LocalFile = 1; // Leave the file locally
    Stream = 2;    // Stream the file to the user directly.
    GoFile = 3;    // See: https://gofile.io/
}

// Request to start a new ROM build
// config_name: Name of the build configuration to use
// target_device: Device codename (e.g., "sailfish")
// rom_name: Name of the ROM (e.g., "LineageOS")
// rom_android_version: Android version (e.g., 11, 12, 12.1)
// build_variant: Build variant (User, UserDebug, Eng)
// parallel_jobs: Number of parallel jobs for the build (optional), defaults to system CPU cores if not set.
// github_token: GitHub token for accessing private repositories (optional)
// upload_method: Method to upload the built ROM artifact
message BuildRequest {
    string config_name = 1;
    string target_device = 2;
    string rom_name = 3;
    float rom_android_version = 4;
    BuildVariant build_variant = 5;
    optional int32 parallel_jobs = 6;
    optional string github_token = 8; // For private repo access
    UploadMethod upload_method = 9;
}

// Result of a build operation
// success: Whether the build was successful
// upload_method: Method used to upload the artifact
// result_details: Details of the result based on upload_method
//   - local_file_path: Path to the local file if upload_method is LocalFile
//   - stream_data: Byte stream of the file if upload_method is Stream (Sent multiple times for large files)
//   - gofile_link: Link to the GoFile upload if upload_method is GoFile
message BuildResult {
    bool success = 1;
    UploadMethod upload_method = 2;
    oneof result_details {
        string local_file_path = 3; // If upload_method is LocalFile
        bytes stream_data = 4;      // If upload_method is Stream (Sent multiple times for large files)
        string gofile_link = 5;     // If upload_method is GoFile
    }
}

enum LogLevel {
    Debug = 0;
    Info = 1;
    Warning = 2;
    Error = 3;
    Fatal = 4;
}

// Log entry for build process
// message: Log message
// timestamp: Unix timestamp of the log entry
// is_finished: Whether this log entry indicates the build has finished
message BuildLogEntry {
    int64 timestamp = 1;
    LogLevel level = 2;
    string message = 3;
    bool is_finished = 4;
}

// Action referencing a specific build by ID
message BuildAction {
    string build_id = 1; // Used for logs, status, cancellation
}

// Response for build submission
// build_id: Unique ID for the submitted build
// accepted: Whether the build request was accepted
// status_message: Additional status information
message BuildSubmission {
    string build_id = 1; 
    bool accepted = 2;
    string status_message = 3;
}

message DirectoryExistsResponse {
    bool exists = 1;
}

// Service for building Android ROMs
service ROMBuildService {
    // Get or set build settings
    rpc GetSettings(google.protobuf.Empty) returns (Settings);
    rpc SetSettings(Settings) returns (google.protobuf.Empty);

    // Clean a specified directory
    rpc CleanDirectory(CleanDirectoryRequest) returns (google.protobuf.Empty);
    rpc DirectoryExists(CleanDirectoryRequest) returns (DirectoryExistsResponse);
    
    // Start a new ROM build.
    rpc StartBuild(BuildRequest) returns (BuildSubmission);

    // Logs streaming for a build in progress
    rpc StreamLogs(BuildAction) returns (stream BuildLogEntry);

    // Cancel a build in progress
    rpc CancelBuild(BuildAction) returns (google.protobuf.Empty);

    // Get the status of a build by ID
    rpc GetStatus(BuildAction) returns (BuildSubmission);

    // Get the result of a completed build
    rpc GetBuildResult(BuildAction) returns (stream BuildResult);
}